### ISSUE TITLE
Maybe something inappropriate in limiting the range of tx or ty ?

### ISSUE THREAD

**Issue by FanvelChen** [jeong-tae/RACNN-pytorch#6]
Hello, in Line 88 of  ./models/RACNN.py
`tx = tx if (in_size/3*2) < tx else (in_size/3*2)`
maybe shoud be
`tx = tx if (in_size/3*2) > tx else (in_size/3*2)`
and the same as that of the ty?

by the way, Would you mind explain the meaning of the Line 115 of ./models/RACNN.py about the backward()  in brief?  I got confused on it .Thank you!
`norm = -(grad_output * grad_output).sum(dim=1)`

**Comment by jeong-tae** [comment]
i found that if cropped image size is too small performance is dropped. so i manually change the cropping rule bcuz i do not have pretrain APN.

For optimization, see “The unreasonable effec- tiveness of noisy data for fine-grained recognition” this paper. i also dont know about detail well. In the RACNN paper, their intention is to keep a consistent direction of optimization.

**Comment by FanvelChen** [comment]
I can got the motivation of your manual change, but what I mean is the direction of comparison maybe is incorrect.
`tx = tx if tx > (in_size/3) else in_size/3`
`tx = tx if (in_size/3*2) < tx else (in_size/3*2)`
`ty = ty if ty > (in_size/3) else in_size/3`
`ty = ty if (in_size/3*2) < ty else (in_size/3*2)`
in my understanding it should be
`tx = tx if tx > (in_size/3) else in_size/3`
`tx = tx if (in_size/3*2) > tx else (in_size/3*2)`
`ty = ty if ty > (in_size/3) else in_size/3`
`ty = ty if (in_size/3*2) > ty else (in_size/3*2)`

Another part is in line 101 in ./models/RACNN.py
`xatt_cropped = xatt[:, h_off : h_end, w_off : w_end]`
should be 
`xatt_cropped = xatt[:, w_off : w_end, h_off : h_end]`
because your x is vertically increasing, and your y is horizontally increasing. According to the corresponding index relation, I think the slice should be changed like this.

**Comment by jeong-tae** [comment]
oh i see. i misunderstand your first question. i will check inquality direction. 

second one is, as far as i know, the index order maybe startwith height and then width so that slice was set as 101 line. i will check again

**Comment by zjuPeco** [comment]
> I can got the motivation of your manual change, but what I mean is the direction of comparison maybe is incorrect.
> `tx = tx if tx > (in_size/3) else in_size/3`
> `tx = tx if (in_size/3*2) < tx else (in_size/3*2)`
> `ty = ty if ty > (in_size/3) else in_size/3`
> `ty = ty if (in_size/3*2) < ty else (in_size/3*2)`
> in my understanding it should be
> `tx = tx if tx > (in_size/3) else in_size/3`
> `tx = tx if (in_size/3*2) > tx else (in_size/3*2)`
> `ty = ty if ty > (in_size/3) else in_size/3`
> `ty = ty if (in_size/3*2) > ty else (in_size/3*2)`
> 
> Another part is in line 101 in ./models/RACNN.py
> `xatt_cropped = xatt[:, h_off : h_end, w_off : w_end]`
> should be
> `xatt_cropped = xatt[:, w_off : w_end, h_off : h_end]`
> because your x is vertically increasing, and your y is horizontally increasing. According to the corresponding index relation, I think the slice should be changed like this.

the index order is (C, H, W).
change line 78-79 in ./models/RACNN.py 
```
x = torch.stack([unit.t()] * 3)
y = torch.stack([unit] * 3)
```
to
```
x = torch.stack([unit] * 3) 
y = torch.stack([unit.t()] * 3)
```
works for me.
because
`mk = (h(x-w_off) - h(x-w_end)) * (h(y-h_off) - h(y-h_end))`

**Comment by libing125** [comment]
> > I can got the motivation of your manual change, but what I mean is the direction of comparison maybe is incorrect.
> > `tx = tx if tx > (in_size/3) else in_size/3`
> > `tx = tx if (in_size/3*2) < tx else (in_size/3*2)`
> > `ty = ty if ty > (in_size/3) else in_size/3`
> > `ty = ty if (in_size/3*2) < ty else (in_size/3*2)`
> > in my understanding it should be
> > `tx = tx if tx > (in_size/3) else in_size/3`
> > `tx = tx if (in_size/3*2) > tx else (in_size/3*2)`
> > `ty = ty if ty > (in_size/3) else in_size/3`
> > `ty = ty if (in_size/3*2) > ty else (in_size/3*2)`
> > Another part is in line 101 in ./models/RACNN.py
> > `xatt_cropped = xatt[:, h_off : h_end, w_off : w_end]`
> > should be
> > `xatt_cropped = xatt[:, w_off : w_end, h_off : h_end]`
> > because your x is vertically increasing, and your y is horizontally increasing. According to the corresponding index relation, I think the slice should be changed like this.
> 
> the index order is (C, H, W).
> change line 78-79 in ./models/RACNN.py
> 
> ```
> x = torch.stack([unit.t()] * 3)
> y = torch.stack([unit] * 3)
> ```
> 
> to
> 
> ```
> x = torch.stack([unit] * 3) 
> y = torch.stack([unit.t()] * 3)
> ```
> 
> works for me.
> because
> `mk = (h(x-w_off) - h(x-w_end)) * (h(y-h_off) - h(y-h_end))`

I think you are right. How is your result? Can you reproduce the result of the paper?

**Comment by libing125** [comment]
https://github.com/jeong-tae/RACNN-pytorch/blob/35716d5e3523fcb371ffb9f7640c6f353ec4cde6/models/RACNN.py#L121
Line 121 should also be changed. Remove the ".t()".

**Comment by YananGu** [comment]
@libing125  Hi, did you reproduce the result of the paper by this code?
